src/Parsing/Test/FStringTests.cs

```diff
@@ -0,0 +1,368 @@
+            }
+        }
+
+        [TestMethod, Priority(0)]
+        public void FStringInternalSubString() {
+            var parser = Parser.CreateParser(MakeReader("f'bla'"), PythonLanguageVersion.V36);
+            var ast = parser.ParseFile();
+            ast.Walk(new MyPythonWalker(expr => {
+                switch (expr) {
+                    case ConstantExpression constExpr:
+                        if (!constExpr.Value.Equals("bla")) {
+                            throw new InternalTestFailureException("Internal const expr didn't match substring");
+                        }
+                        break;
+                }
+            }));
+        }
+
+        [TestMethod, Priority(0)]
+        public void CanEscapeBraces() {
+            var parser = Parser.CreateParser(MakeReader("f'{{bla}}'"), PythonLanguageVersion.V36);
+            var ast = parser.ParseFile();
+            ast.Walk(new MyPythonWalker(expr => {
+                switch (expr) {
+                    case ConstantExpression constExpr:
+                        if (!constExpr.Value.Equals("{bla}")) {
+                            throw new InternalTestFailureException("Internal const expr didn't match substring");
+                        }
+                        break;
+                }
+            }));
+        }
+
+        [TestMethod, Priority(0)]
+        public void FStringWithExpressionInside() {
+            var parser = Parser.CreateParser(MakeReader("f'{x}'"), PythonLanguageVersion.V36);
+            var ast = parser.ParseFile();
+            var found = false;
+            ast.Walk(new MyPythonWalker(expr => {
+                switch (expr) {
+                    case FStringExpression fStringExpr:
+                        if (fStringExpr.GetChildNodes().First() is NameExpression xVarExpr) {
+                            xVarExpr.Name.Equals("x").Should().BeTrue();
+                            found = true;
+                        }
+                        break;
+                }
+            }));
+            found.Should().BeTrue();
+        }
+
+        [TestMethod, Priority(0)]
+        public void FStringWithStringInside() {
+            char doubleQuote = '"';
+            var number = 111;
+            var code = $@"
+            s = f'f{doubleQuote}{number}{doubleQuote}'
+            ";
+            var parser = Parser.CreateParser(MakeReader(code), PythonLanguageVersion.V36);
+            var ast = parser.ParseFile();
+            ast.Walk(new MyPythonWalker(expr => {
+                switch (expr) {
+                    case FStringExpression outerFstring:
+                        if (outerFstring.GetChildNodes().First() is FStringExpression innerFString) {
+                            if (outerFstring.GetChildNodes().First() is ConstantExpression constExpr) {
+                                constExpr.Value.Should().BeEquivalentTo("f\"111\"");
+                            }
```

Comment:
The test seems to only verify when `case FStringExpression outerFstring`. If condition is false, the test will pass which doesn't seem right.

Reply:
The best for parser testing is to have some sort of AST persistence to disk so output of the test can be compared to a predefined baseline. Ex https://github.com/Microsoft/RTVS/blob/master/src/R/Core/Test/Parser/ParseBuiltInsTest.cs